FlatterScatter

chillenzer opened the pull request:
This PR adds amongst a number of modernisations the new FlatterScatter creation policy.

chillenzer committed:
Update allocator.hpp

chillenzer committed:
Update device_allocator.hpp

chillenzer committed:
Update old creation policies

chillenzer committed:
Update XMallocSIMD

chillenzer committed:
Update utils

Commit:
Run clang-format

chillenzer committed:
Add FlatterScatter

chillenzer committed:
Update LICENSE, README and Usage

chillenzer committed:
Update examples

chillenzer committed:
Update tests

chillenzer committed:
Add catch2 as submodule

chillenzer committed:
Update CMakeLists.txt including version bump for Catch2

chillenzer committed:
Minor adjustments to CI

chillenzer committed:
Update license headers

psychocoderHPC commented on examples/mallocMC_example01.cpp:
```diff
@@ -26,90 +27,98 @@
-    mallocMC::ReservePoolPolicies::AlpakaBuf<Acc>,
-    mallocMC::AlignmentPolicies::Shrink<ShrinkConfig>>;
-
 ALPAKA_STATIC_ACC_MEM_GLOBAL int** arA;
```
IMO this must be equal to https://github.com/alpaka-group/alpaka/blob/7e1b6ce24c5fadc9c58dd0d1dc59c022e96a087b/test/unit/mem/view/src/ViewStaticAccMem.cpp#L19

`alpaka::DevGlobal` is missing and the usage requires a `.get()` but it looks it runs and the `DevGlobal` is only required for SYCL. SO I am fine to keep it as it is.

chillenzer replied:
Opened: https://github.com/alpaka-group/mallocMC/issues/272 to track this.

psychocoderHPC commented on src/include/mallocMC/creationPolicies/FlatterScatter.hpp:
```diff
@@ -0,0 +1,456 @@
+            using VecType = alpaka::Vec<Dim, Idx>;
+
+            auto poolView = alpaka::createView(dev, reinterpret_cast<char*>(pool), alpaka::Vec<Dim, Idx>(memsize));
+            alpaka::memset(queue, poolView, 0U);
```
note: You mentioned last time offline that we use a kernel to initialize the heap but here the memset is still present. It is fine for this PR I would like to avoid that we use memset and a on device initialization.

chillenzer replied:
Oh, good catch! There is an `init` method for the heap but apparently it is unused. It's on my TODO list.

chillenzer replied:
Opened https://github.com/alpaka-group/mallocMC/issues/273 to track this.

psychocoderHPC commented on src/include/mallocMC/creationPolicies/FlatterScatter.hpp:
```diff
@@ -0,0 +1,456 @@
+
+            auto workDivSingleThread
+                = alpaka::WorkDivMembers<Dim, Idx>{VecType::ones(), VecType::ones(), VecType::ones()};
+            alpaka::exec<TAcc>(queue, workDivSingleThread, FlatterScatterAlloc::InitKernel{}, heap, pool, memsize);
```
This sounds like the kernel which is performing the on device init equal to the memset above.

chillenzer replied:
Oh, damn! It's much worse! The above memset is completely redundant. -.-

psychocoderHPC commented on src/include/mallocMC/creationPolicies/FlatterScatter/AccessBlock.hpp:
```diff
@@ -0,0 +1,823 @@
+        ALPAKA_FN_INLINE ALPAKA_FN_ACC auto getAvailableMultiPages(auto const& /*acc*/, uint32_t const chunkSize) const
+            -> uint32_t
+        {
+            // TODO(lenz): This is not thread-safe!
```
we should not forget this todo and handle it in a future PR

chillenzer replied:
This is kind of a minor issue: The informational methods `getAvailableSlots` and its helpers provide information that cannot possibly be used in a thread-safe manner. So, calling them while other threads are write-accessing the same allocator is not a reasonable thing to do anyways. Mostly left it in for informational purposes but maybe I should drop the `TODO(lenz)` part. It is documented in the `Usage.md` document by the way.

psychocoderHPC commented on src/include/mallocMC/creationPolicies/FlatterScatter/AccessBlock.hpp:
```diff
@@ -0,0 +1,823 @@
+            TAcc const& acc,
+            uint32_t const index,
+            uint32_t const numBytes,
+            uint32_t& chunkSizeCache) -> bool
```
only a note: this should be refactored in the future and return the value together with the current return value as tuple.

chillenzer replied:
Opened https://github.com/alpaka-group/mallocMC/issues/274 to track this.

psychocoderHPC reviewed (changes requested):
offline discussed catch2 should be fetched on the need to not increase the repository size to much.

The rest looks good! ONly one small question for the heap init part is open.

chillenzer committed:
Remove redundant memset

chillenzer committed:
Add submodule to CI

chillenzer commented:
Okay. I've surgically removed the Catch2 commit and replaced it with a git submodule. I've removed the redundant `memset` in the init method and will certainly do the initialisation benchmark again! Other than that I've opened issues to track your good suggestions.
